pkg/placement/placement.go

```diff
@@ -72,220 +128,57 @@ func (p *Service) ReportDaprStatus(srv placementv1pb.Placement_ReportDaprStatusS
-					placementOptions{incrementGeneration: false})
-				log.Debugf("New member is added: %s", registeredMemberID)
-				monitoring.RecordHostsCount(len(p.hosts))
+				p.addRuntimeConnection(ctx, srv)
+				p.performTablesUpdate([]placementGRPCStream{srv}, p.raftNode.FSM().PlacementState())
+				log.Debugf("Stream connection is establiched from %s", registeredMemberID)
 			}
 
-			_, err := p.raftNode.ApplyCommand(raft.MemberUpsert, raft.DaprHostMember{
-				Name:     req.Name,
-				AppID:    req.Id,
-				Entities: req.Entities,
-			})
-			if err != nil {
-				log.Debugf("fail to apply command: %v", err)
+			p.membershipCh <- hostMemberChange{
+				cmdType: raft.MemberUpsert,
+				host: raft.DaprHostMember{
+					Name:     req.Name,
+					AppID:    req.Id,
+					Entities: req.Entities,
+				},
 			}
 
-			p.ProcessHost(req)
-
 		default:
 			if registeredMemberID == "" {
-				log.Debug("stream is disconnected before member is added")
+				log.Error("stream is disconnected before member is added")
 				return nil
 			}
 
-			p.hostsLock.Lock()
-			p.RemoveHost(srv)
-
-			_, err := p.raftNode.ApplyCommand(raft.MemberRemove, raft.DaprHostMember{
-				Name: registeredMemberID,
-			})
-
-			if err != nil {
-				log.Debugf("fail to apply command: %v", err)
-			}
-
-			// TODO: Need the robust fail-over handling by the intermittent
-			// network outage to prevent from rebalancing actors.
-			p.ProcessRemovedHost(registeredMemberID)
-			p.hostsLock.Unlock()
-
+			p.deleteRuntimeConnection(srv)
 			if err == io.EOF {
-				log.Debugf("Member is removed gracefully: %s", registeredMemberID)
+				log.Debugf("Stream connection is disconnected gracefully: %s", registeredMemberID)
+				p.membershipCh <- hostMemberChange{
+					cmdType: raft.MemberRemove,
+					host:    raft.DaprHostMember{Name: registeredMemberID},
+				}
 			} else {
-				log.Debugf("Member is removed with error: %s, %v", registeredMemberID, err)
+				// no actions for hashing table. Instead, MembershipChangeWorker will check
+				// host updatedAt and if now - updatedAt > faultyHostDetectMaxDuration, remove hosts.
+				log.Debugf("Stream connection is disconnected with the error: %v", err)
 			}
 
 			return nil
 		}
 	}
 }
 
-func (p *Service) addHost(ctx context.Context, srv placementv1pb.Placement_ReportDaprStatusServer) {
-	p.hostsLock.Lock()
-	defer p.hostsLock.Unlock()
-
-	p.hosts = append(p.hosts, srv)
-}
-
-// RemoveHost removes the host from the hosts list
-func (p *Service) RemoveHost(srv placementv1pb.Placement_ReportDaprStatusServer) {
-	for i := len(p.hosts) - 1; i >= 0; i-- {
-		if p.hosts[i] == srv {
-			p.hosts = append(p.hosts[:i], p.hosts[i+1:]...)
-		}
-	}
-}
-
-// PerformTablesUpdate updates the connected dapr runtimes using a 3 stage commit. first it locks so no further dapr can be taken
-// it then proceeds to update and then unlock once all runtimes have been updated
-func (p *Service) PerformTablesUpdate(hosts []placementv1pb.Placement_ReportDaprStatusServer,
```

Comment:
this function is refactored to remove all duplicated codes in membership.go. I added unit-tests to validate lock/update/unlock stages.